File: 0146-lru-cache/0146-lru-cache.py
class Node:
    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.next = None
        self.prev = None

class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity 
        self.head = Node(-1, -1)
        self.tail = Node(-1, -1)
        self.head.next = self.tail
        self.tail.prev = self.head
        self.cacheMap = collections.defaultdict(Node)

    def get(self, key: int) -> int:
        if key in self.cacheMap:
            node = self.cacheMap[key]
            self.removeNode(node)
            self.addNodeToHead(node)
            self.cacheMap[key] = node
            return node.value
        else:
            return -1

    def put(self, key: int, value: int) -> None:
        if key in self.cacheMap:
            self.removeNode(self.cacheMap[key])
            del self.cacheMap[key]

        newNode = Node(key, value)            
        self.addNodeToHead(newNode)
        self.cacheMap[key] = newNode

        if len(self.cacheMap) > self.capacity:
            prevNode = self.head.next
            self.removeNode(prevNode)
            del self.cacheMap[prevNode.key]

    def removeNode(self, node):
        prevNode = node.prev
        nextNode = node.next
        prevNode.next = nextNode
        nextNode.prev = prevNode

    def addNodeToHead(self, node):
        prevNode = self.tail.prev
        prevNode.next = node
        self.tail.prev = node
        node.next = self.tail
        node.prev = prevNode

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 0146-lru-cache/0146-lru-cache.java
class Node {

    int key;
    int value;
    Node next;
    Node prev;

    public Node(int key, int value){
        this.key = key;
        this.value = value;
    }
}

class LRUCache {

    int capacity;
    Map<Integer,Node> map = new HashMap<>();
    Node head = new Node(-1, -1);
    Node tail = new Node(-1, -1);

    public LRUCache(int capacity) {
        this.capacity = capacity;        
        head.next = tail;
        tail.prev = head;
    }
    
    public int get(int key) {
        if (map.containsKey(key)){
            Node preNode = map.get(key);
            removeNode(preNode);
            addToHead(preNode);
            map.put(key, preNode);            
            return map.get(key).value;
        }
        
        return -1;
    }
    
    public void put(int key, int value) {
        if(map.containsKey(key)){
            Node preNode = map.get(key);
            removeNode(preNode);
            map.remove(key);
        }

        Node newNode = new Node(key, value);
        addToHead(newNode);
        map.put(key, newNode);

        if (map.size() > this.capacity){
            Node preNode = tail.prev;
            removeNode(preNode);
            map.remove(preNode.key);
        }
    }

    public void removeNode(Node target){
        Node nextNode = target.next;
        Node prevNode = target.prev;
        prevNode.next = nextNode;
        nextNode.prev = prevNode;
    }

    public void addToHead(Node newNode){
        Node temp = head.next;
        head.next = newNode;
        newNode.prev = head;
        newNode.next = temp;
        temp.prev = newNode;
    }    
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */